presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1424,6 +1435,55 @@ else if (materializedViewStatus.isPartiallyMaterialized()) {
             return SqlFormatterUtil.getFormattedSql(unionQuery, sqlParser, Optional.empty());
         }
 
+        /**
+         * Returns materialized view status, considering filter conditions from query
+         */
+        private MaterializedViewStatus getMaterializedViewStatus(QualifiedObjectName materializedViewName, Optional<Scope> scope)
+        {
+            TupleDomain<String> baseQueryDomain;
+            // We can use filter from base query WHERE clause to consider only relevant partitions when getting
+            // materialized view status
+            // TODO: Collect predicates in outer queries and apply the union to this criteria. i.e. in
+            //  SELECT x FROM (SELECT x FROM tbl WHERE y < 20) WHERE y > 5 where y is a partition key,
+            //  right now we would only remove partitions y >= 20. We should eventually consider y <= 5 as well.
+            if (currentSubquery.getWhere().isPresent()) {
+                Scope sourceScope = analyzeFrom(currentSubquery, scope);
+                Expression baseQueryWhereClause = currentSubquery.getWhere().get();
+
+                analyzeWhere(currentSubquery, sourceScope, baseQueryWhereClause);
+
+                DomainTranslator domainTranslator = new RowExpressionDomainTranslator(metadata);
+                RowExpression rowExpression = SqlToRowExpressionTranslator.translate(
+                        baseQueryWhereClause,
+                        analysis.getTypes(),
+                        ImmutableMap.of(),
+                        metadata.getFunctionAndTypeManager(),
+                        session);
+
+                try {
+                    DomainTranslator.ExtractionResult<String> predicateFromBaseQuery = domainTranslator.fromPredicate(
+                            session.toConnectorSession(),
+                            rowExpression,
+                            (baseFilterExpression, domain) -> baseFilterExpression instanceof VariableReferenceExpression ?
+                                    Optional.of(((VariableReferenceExpression) baseFilterExpression).getName()) :
+                                    Optional.empty());
+                    baseQueryDomain = predicateFromBaseQuery.getTupleDomain();
+                }
+                // Sometimes IllegalStateExceptions are thrown due to type mismatch between types like bigint and integer, which should
+                // not be happening. https://github.com/prestodb/presto/issues/18034
+                // instead of failing optimization outright when we cannot determine domain of filter condition, simply do not consider the
+                // filter condition for purposes of deciding if query has too many missing partitions.
+                catch (Exception e) {
```

Comment:
Don't catch all exceptions. Also we have this code in 2 places. Can we create a common function which does this try catch so the logic is isolated in one place